barman/config.py

```diff
@@ -1314,6 +1320,228 @@ def _validate_with_keys(config_items, allowed_keys, section):
+        self.key = key
+        self.value = value
+        self.config_file = config_file
+
+    def __eq__(self, other):
+        """Equality support."""
+        if isinstance(other, ConfigChange):
+            return self.as_tuple() == other.as_tuple()
+        return False
+
+    def __hash__(self):
+        """Hash/set support."""
+        return hash(self.as_tuple())
+
+    def as_tuple(self) -> tuple:
+        """Convert to a tuple, ordered as self._fields."""
+        return tuple(vars(self)[k] for k in self._fields)
+
+    def as_dict(self):
+        return {k: vars(self)[k] for k in self._fields}
+
+    @classmethod
+    def from_dict(cls, obj):
+        """
+        Factory for configuration change objects.
+
+        Generates configuration change objects starting from a dictionary with
+        the same fields.
+
+        """
+        if set(obj.keys()) == set(cls._fields):
+            return cls(**obj)
+        raise ValueError("Malformed configuration change serialization: %r" % obj)
+
+
+class ConfigChangeSet:
+    _fields = ["section", "changes_set"]
+
+    def __init__(self, section, changes_set=[]):
+        self.section = section
+        self.changes_set = changes_set
+
+    def __eq__(self, other):
+        """Equality support."""
+        if isinstance(other, ConfigChangeSet):
+            return self.as_tuple() == other.as_tuple()
+        return False
+
+    def __hash__(self):
+        """Hash/set support."""
+        return hash(self.as_tuple())
+
+    def as_tuple(self):
+        """Convert to a tuple, ordered as self._fields."""
+        return tuple(vars(self)[k] for k in self._fields)
+
+    def as_dict(self):
+        return {k: vars(self)[k] for k in self._fields}
+
+    @classmethod
+    def from_dict(cls, obj):
+        """
+        Factory for configuration change objects.
+
+        Generates configuration change objects starting from a dictionary with
+        the same fields.
+
+        """
+        if set(obj.keys()) == set(cls._fields):
+            return cls(**obj)
+        if set(obj.keys()) == set(ConfigChange._fields):
+            return ConfigChange(**obj)
+        raise ValueError("Malformed configuration change serialization: %r" % obj)
+
+
+class ConfigChangesQueue:
+    """
+    Wraps the management of the config changes queue.
+
+    Once instantiated the queue can be accessed using the `queue` property.
+    """
+
+    def __init__(self, queue_file):
+        self.queue_file = queue_file
+        self._queue = None
+        self.open()
+
+    @staticmethod
+    def read_file(path) -> List[ConfigChangeSet]:
+        """Reads a json file containing a list of configuration changes."""
+        try:
+            with open(path, "r") as queue_file:
+                # Read the queue if exists
+                return json.load(queue_file, object_hook=ConfigChangeSet.from_dict)
+        except FileNotFoundError:
+            return []
+
+    def __enter__(self):
+        return self
```

Comment:
Maybe we should call `self.open()` here instead of on `__init__`?

Reply:
init ensure the loading at creation class, enter execution happens later in the sequence of actions...
tbh the sooner we load it the better...

Reply:
I see. The sooner we can detect issues too in that case.